server/src/main/java/org/elasticsearch/rest/RestUtils.java

```diff
@@ -146,17 +146,16 @@ private static String decodeComponent(final String s, final Charset charset, boo
     }
 
     private static boolean decodingNeeded(String s, int size, boolean plusAsSpace) {
-        boolean decodingNeeded = false;
-        for (int i = 0; i < size; i++) {
+        if (Strings.isEmpty(s)) {
+            return false;
+        }
+        for (int i = 0; i < Math.min(s.length(), size); i++) {
```

Comment:
Just for this function, there is no guarantee that the parameter size will not cross the boundary during the loop. Of course, since it is a private method, it can be checked when it is called externally. Personally, I prefer to ensure the integrity of the method logic.

Reply:
Likewise, parameter `s` may be null. :)

Reply:
@shane-zfx I don't quite understand your argument. What is the downside of having
```
int num = Math.min(s.length(), size);
for (int i=0; i<num; i++) {

}
```
?

Reply:
Yes, I missed this part.